FEAT: Sink constant offsets down a GEP chain to tail for reduction of register usage.

StevenYangCC opened the pull request:
Sink constant offset in a GEP chain to tail. For example,
%gep0 = getelementptr half, ptr addrspace(3) %ptr, i32 512
%gep1 = getelementptr half, ptr addrspace(3) %gep0, i32 %ofst0
%gep2 = getelementptr half, ptr addrspace(3) %gep1, i32 %ofst1
%data = load half, ptr addrspace(3) %gep2, align 2
==>
%gep0 = getelementptr half, ptr addrspace(3) %ptr, i32 %ofst0
%gep1 = getelementptr half, ptr addrspace(3) %gep0, i32 %ofst1
%gep2 = getelementptr half, ptr addrspace(3) %gep1, i32 512
%data = load half, ptr addrspace(3) %gep2, align 2

arsenm commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1344,6 +1362,133 @@ bool SeparateConstOffsetFromGEP::reuniteExts(Function &F) {
+      NewResTy = ResTy;
+      int64_t NewIdxValue = cast<ConstantInt>(BaseIdx)->getSExtValue() +
+                            cast<ConstantInt>(Idx)->getSExtValue();
+      Type *NewIdxType = (NewIdxValue < std::numeric_limits<int32_t>::min() ||
+                          NewIdxValue > std::numeric_limits<int32_t>::max())
+                             ? Type::getInt64Ty(GEP->getContext())
+                             : Type::getInt32Ty(GEP->getContext());
+      NewIdx = ConstantInt::get(NewIdxType, NewIdxValue);
```
The type should just be the correct offset type for the pointer

arsenm commented on llvm/test/CodeGen/PowerPC/split-index-tc.ll:
```diff
@@ -77,5 +77,3 @@ if.end89:                                         ; preds = %if.else59
 }
 
 attributes #0 = { nounwind }
-
-
```
Whitespace only change

arsenm commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1344,6 +1362,133 @@ bool SeparateConstOffsetFromGEP::reuniteExts(Function &F) {
+bool SeparateConstOffsetFromGEP::sinkGEPConstantOffset(Function &F) {
+  bool Changed = false;
+  SmallVector<Value *, 4> Candidates;
+  for (BasicBlock &B : F) {
```
Shouldn't need a separate walk over the function? As-is this is a separate pass that happens to exist in SeparateConstOffsetFromGEP

arsenm reviewed (commented):
Needs new tests that show the transformation (preferably precommitted so the diff easily shows what this did)

StevenYangCC commented:
> Needs new tests that show the transformation (preferably precommitted so the diff easily shows what this did)

@john Thanks for catching these! I have fixed this, could you please verify the change?

StevenYangCC commented:
@john Thanks for catching these! I have fixed this, could you please verify the change?

StevenYangCC commented:
/retry-label

StevenYangCC committed:
[SinkGEPConstOffset] FEAT: Sink constant offsets down a GEP chain to tail for reduction of register usage.

Summary:

Sink constant offsets down the GEP chain to the tail helps reduce register usage.
For example:

%gep0 = getelementptr half, ptr addrspace(3) %ptr, i32 512
%gep1 = getelementptr half, ptr addrspace(3) %gep0, i32 %ofst0
%gep2 = getelementptr half, ptr addrspace(3) %gep1, i32 %ofst1
%data = load half, ptr addrspace(3) %gep2, align 2

==>

%gep0 = getelementptr half, ptr addrspace(3) %ptr, i32 %ofst0
%gep1 = getelementptr half, ptr addrspace(3) %gep0, i32 %ofst1
%gep2 = getelementptr half, ptr addrspace(3) %gep1, i32 512
%data = load half, ptr addrspace(3) %gep2, align 2

StevenYangCC commented:
@arsenm @evanphx @haberman Could you please review the code

StevenYangCC committed:
Merge branch 'main' into feat/sink-gep-constant-offset
